shm_wrapper_aux/shm_wrapper_aux.c

```diff
@@ -0,0 +1,469 @@
+			break;
+		}
+		usleep(1000);
+		gettimeofday(&curr, NULL);
+		time_taken = (double)(curr.tv_sec - start.tv_sec) + ((double)(curr.tv_usec - start.tv_usec) / 1000000.0);
+	}
+	if (time_taken >= DESC_WR_MAXBLK) {
+		sprintf(msg, "%s overwriting field %d to value %d", (process == NET_HANDLER)? "NET_HANDLER" : "EXECUTOR", field, val);
+		log_runtime(WARN, msg);
+	}
+	
+	//wait on rd_sem
+	if (sem_wait(rd_sem) == -1) {
+		error("sem_wait: robot_desc_mutex");
+	}
+	
+	//write the val into the field, and set appropriate pending element to 1
+	rd_ptr->fields[field] = val;
+	if (process == NET_HANDLER) {
+		rd_ptr->net_to_exe_pending = 1;
+	} else if (process == EXECUTOR) {
+		rd_ptr->exe_to_net_pending = 1;
+	}
+	
+	//release rd_sem
+	if (sem_post(rd_sem) == -1) {
+		error("sem_post: robot_desc_mutex");
+	}
+}
+
+/*
+This function reads the current state of the gamepad to the provided pointers.
+Blocks on both the gamepad semaphore and device description semaphore (to check if gamepad connected).
+	- pressed_buttons: pointer to 32-bit bitmap to which the current button bitmap state will be read into
+	- joystick_vals: array of 4 floats to which the current joystick states will be read into
+No return value.
+*/
+void gamepad_read (process_t process, uint32_t *pressed_buttons, float *joystick_vals)
+{
+	//wait on rd_sem
+	if (sem_wait(rd_sem) == -1) {
+		error("sem_wait: robot_desc_mutex");
+	}
+	
+	//if no gamepad connected, then release rd_sem and return
+	if (rd_ptr->fields[GAMEPAD] == DISCONNECTED) {
+		log_runtime(WARN, "tried to read, but no gamepad connected");
+		if (sem_post(rd_sem) == -1) {
+			error("sem_post: robot_desc_mutex");
+		}
+		return;
+	}
+	
+	//release rd_sem
+	if (sem_post(rd_sem) == -1) {
+		error("sem_post: robot_desc_mutex");
+	}
+	
+	//wait on gp_sem
+	if (sem_wait(gp_sem) == -1) {
+		error("sem_wait: gamepad_mutex");
+	}
+	
+	*pressed_buttons = gp_ptr->buttons;
+	for (int i = 0; i < 4; i++) {
+		joystick_vals[i] = gp_ptr->joysticks[i];
+	}
+	
+	//release gp_sem
+	if (sem_post(gp_sem) == -1) {
+		error("sem_post: gamepad_mutex");
+	}
+}
+
+/*
+This function writes the given state of the gamepad to shared memory.
+Blocks on both the gamepad semaphore and device description semaphore (to check if gamepad connected).
+	- pressed_buttons: a 32-bit bitmap that corresponds to which buttons are currently pressed
+		(only the first NUM_GAMEPAD_BUTTONS bits used, since there are NUM_GAMEPAD_BUTTONS buttons)
+	- joystick_vals: array of 4 floats that contain the values to write to the joystick
+No return value.
+*/
+void gamepad_write (process_t process, uint32_t pressed_buttons, float *joystick_vals)
+{
+	//wait on rd_sem
+	if (sem_wait(rd_sem) == -1) {
+		error("sem_wait: robot_desc_mutex");
+	}
+	
+	//if no gamepad connected, then release rd_sem and return
+	if (rd_ptr->fields[GAMEPAD] == DISCONNECTED) {
+		log_runtime(WARN, "tried to write, but no gamepad connected");
```

Comment:
Like above, think this should be an error.

Reply:
Done